### PR TITLE
[Improvement](dict) optimize dictionary column

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

I tested this on ckbench q25 and print flame graph.
Baseline
image

Optimized
image

We got overall 1.8X performance boost on q25 

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/columns/column_dictionary.h [hunk 1]
```diff
@@ -360,40 +362,58 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
...
+                bool operator()(const size_t a, const size_t b) const {
+                    return _comparator(_dict_data[a], _dict_data[b]);
+                }
+
+            private:
+                StringValue::Comparator _comparator;
+                DictContainer& _dict_data;
+            };
+            Comparator comparator(*_dict_data);
+            std::sort(_perm.begin(), _perm.end(), comparator);
```
Comment (adonis0147) [inline be/src/vec/columns/column_dictionary.h:409]:
```suggestion
            std::sort(_perm.begin(), _perm.end(),
                      [&dict_data = *_dict_data, &comparator = _comparator](const size_t a,
                                                                            const size_t b) {
                          return comparator(dict_data[a], dict_data[b]);
                      });

```

**Review by adonis0147 — DISMISSED** [review]

File: be/src/vec/columns/column_dictionary.h [hunk 2]
```diff
@@ -192,11 +192,13 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
...
     Status filter_by_selector(const uint16_t* sel, size_t sel_size, IColumn* col_ptr) override {
         auto* res_col = reinterpret_cast<vectorized::ColumnString*>(col_ptr);
+        res_col->get_offsets().reserve(sel_size);
+        res_col->get_chars().reserve(_dict.avg_str_len() * sel_size);
         for (size_t i = 0; i < sel_size; i++) {
             uint16_t n = sel[i];
             auto& code = reinterpret_cast<T&>(_codes[n]);
             auto value = _dict.get_value(code);
-            res_col->insert_data(value.ptr, value.len);
+            res_col->insert_data_without_reserve(value.ptr, value.len);
```
Comment (BiteTheDDDDt) [inline be/src/vec/columns/column_dictionary.h:201]: I think `_dict.avg_str_len() * sel_size` may be less than sum length of elements.

Comment (Gabriel39) [reply be/src/vec/columns/column_dictionary.h:201]: If so, `chars` in ColumnString will still to reserve a bigger memory block. `_dict.avg_str_len() * sel_size` is just a conservative estimation here.

File: be/src/vec/columns/column_dictionary.h (hunk 1 above)
Comment (Gabriel39) [reply be/src/vec/columns/column_dictionary.h]: Done, thanks for your suggestion!

File: be/src/vec/columns/column_dictionary.h [hunk 3]
```diff
@@ -360,40 +356,50 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
...
+                if (values.find((*_dict_data)[i]) != values.end()) {
+                    selected[i] = true;
                 }
             }
         }
 
         void clear() {
-            _dict_data.clear();
-            _inverted_index.clear();
-            _code_convert_table.clear();
```
Comment (zenoyang) [inline be/src/vec/columns/column_dictionary.h:383]: why remove `_code_convert_table.clear()`?
In addition, can you provide the performance improvement data for this optimization?

Comment (Gabriel39) [reply be/src/vec/columns/column_dictionary.h:383]: Sorry, this is a mistake and I fixed it.

Btw, I provide a performance test in this PR's description

**Comment by zenoyang** [comment]
LGTM

**Review by adonis0147 — APPROVED** [review]
LGTM

**Pull request merged**